A lookup button added to the mobile toolbar

kirlat opened the pull request:
Added a lookup button to the mobile UI as per #182. A lookup button will open an action panel with the lookup input only. The tools button will open a lookup panel with tools icons only. The second click on the button will close the previously opened panel: this is the only way to close a panel that was opened. 

If a panel of a different kind (lookup/tools) is open when a toolbar's button was clicked, the newly opened panel will replace a previous one: the previous one will be closed and the new one will be opened instead.

Will remove a branch dependency upon merging.

irina060981 commented on alpheios_nemo_ui/data/templates/main/container.html:
```diff
@@ -79,12 +79,38 @@
             })
           embedded.activate().then(function () {
             if (embedded.platform.isMobile) {
+              let lookupPanelIsVisible = false
```
May be it is better to place it outside? inside some js function - in js file or even inside embed-lib?
To prevent this basic template file to become bigger with the custom code that is not directly needed for templating?
@kirlat , what do you think?

kirlat replied:
I was just trying to follow the pattern set by the previous Nemo UI development.

I was thinking about putting this code into an embed lib. Giving it some thought, I decided that this logic that specifies what kind of panel to open or close when a Nemo UI button is clicked is very Nemo UI specific and thus has no place in the embed-lib.

Putting this code to a separate file would allow this file to be cached but, at the same time, will add an extra request and make code management, on my opinion, a little more complex (i.e the code that will run on activation will be in two places: inlined inside this template and in a separate file). It might be worth putting all embedded JS code from this template into a separate file, but this change will be to large and not related directly to the goal of this PR so I think it's probably not a good option now.

There are some repetitions with the code added, but I've done it intentionally to make it simpler to understand because I think simplicity is sometimes more important than the more efficient, but the trickier code (especially in simple cases like that). To eliminate this repetitiveness, I can split this code into a separate, more abstract function and call it twice. But, as a result of this, the code will become a little more complex (am I being to overcautious here?). What do you think? @balmas, what is your opinion on that?

irina060981 replied:
I could explain why I think that this code should be moved to a separate js file - not all embed-lib activation - but this additional code with some details how exactly it should be used in alpheios-nemo-ui.

1. The embed-lib functionality is only a part of an alpheios-nemo-ui, it is used (from the code perspective) only in one place as an itialization of a outstanding library. But the container.html file is used as a basic template for all other templates. It has much more bocks and initializations inside. And adding so much lines of code here that has no any sense (from the template perspective of course) for defining blocks used inside and that could be used inside children template.

2. If you look at other templates there are more js code than used only for embed-lib - and all js code is packed inside separate libraries in js folder. I believe that it is done to remove additional code from flask templates and remain them to be easy to read.

> Putting this code to a separate file would allow this file to be cached but, at the same time, will add an extra request and make code management, on my opinion, a little more complex 
 
3. I quitely disagree with that the approach of creating smaller files makes code management more complex. I believe that such size of the file as we have for uicontroller - couldn't be properly managed at all. Everytime when I should work with it I dream that it becomes a small and more readable sometimes.

I believe that flask templates should have the main purpose - they should define available and used blocks. And the extending of the embed-lib functionality is out of this purpose.

> There are some repetitions with the code added

From my point of view - I don't think that repetions of the code are so bad - sometimes it is much easier to read. It is much harder for me to understand properly some mordern ES6 constructions - yes, they are pretty - but it is often the single advantage of them.
And if use the approach of smaller files - then such subjective opinions of the code constructions loose there sense. So I have no suggestions from this point of view.

I ask for only to put this code to a separate file and only execute it with on line of the code.

```
if (embedded.platform.isMobile) {
              let lookupPanelIsVisible = false
              let toolsPanelIsVisible = false
              let lookupEl = document.querySelector('#alph-lookup-ctrl')
              if (lookupEl) {
                lookupEl.addEventListener('click', function () {
                  if (toolsPanelIsVisible) {
                    embedded.closeActionPanel()
                    toolsPanelIsVisible = false
                  }
                  if (!lookupPanelIsVisible) {
                    embedded.openActionPanelLookup()
                    lookupPanelIsVisible = true
                  } else {
                    embedded.closeActionPanel()
                    lookupPanelIsVisible = false
                  }
                }, { passive: true })
              }
              let toolsEl = document.querySelector('#alph-tools-ctrl')
              if (toolsEl) {
                toolsEl.addEventListener('click', function () {
                  if (lookupPanelIsVisible) {
                    embedded.closeActionPanel()
                    lookupPanelIsVisible = false
                  }
                  if (!toolsPanelIsVisible) {
                    embedded.openActionPanelToolbar()
                    toolsPanelIsVisible = true
                  } else {
                    embedded.closeActionPanel()
                    toolsPanelIsVisible = false
                  }
                }, { passive: true })
              }
            }
```

kirlat replied:
So are you saying that in a file as small and targeted as `container.html` is such piece of code of after embed-lib initialization will hurt its readability? I think there is a lot of truth in it. But even with that, an extra file will be an extra HTTP request, and that's just for the small piece of code. So I can't really decide 🙂. @balmas, are we planning to move to HTTP/2 for the Alpheios reader? That will make those performance concerns irrelevant.

This piece of code is needed only for those template blocks that include the lookup and tools buttons. So, ideally, we would include it only into pages that embeds the buttons. But the template structure seems pretty convoluted, and I'm not sure there is an easy way to do that 🙁.

balmas replied:
> are we planning to move to HTTP/2 for the Alpheios reader? 
Yes, we will move to HTTP/2 for the reader. It's on my list of things for the release. 

>This piece of code is needed only for those template blocks that include the lookup and tools buttons.
I think that's pretty much every page of the application so having it in the container.html seems appropriate to me.

I think you both have very valid points here. If the move to http/2 will alleviate the performance concern of a separate js file, then I would vote for that.   I will move that higher on my priority list.

kirlat replied:
> Yes, we will move to HTTP/2 for the reader. It's on my list of things for the release.

Great! I will move it into a separate file then.

kirlat commented:
I've moved the code discussed above into a separate file that is loaded dynamically. I've also fixed an issue that was causing the buttons to spill over the toolbar on resolutions 360px and a little over.

irina060981 commented:
Looks good!
